TestUpdateDefaultReplicaConfig is unstable

## Flaky Test

### Which jobs are failing
```
2023-11-22T04:57:42.1796276Z --- FAIL: TestConfigTestSuite (78.16s)
2023-11-22T04:57:42.1796660Z     --- FAIL: TestConfigTestSuite/TestUpdateDefaultReplicaConfig (32.09s)
2023-11-22T04:57:42.1796932Z         testutil.go:264: start to run test in pd mode
2023-11-22T04:57:42.1797201Z         testutil.go:268: finish to run test in pd mode
2023-11-22T04:57:42.1797461Z         testutil.go:273: start to run test in api mode
2023-11-22T04:57:42.1797626Z         testutil.go:64: 
2023-11-22T04:57:42.1798125Z             	Error Trace:	/home/runner/work/pd/pd/pkg/utils/testutil/testutil.go:64
2023-11-22T04:57:42.1798830Z             	            				/home/runner/work/pd/pd/tests/pdctl/config/config_test.go:690
2023-11-22T04:57:42.1799534Z             	            				/home/runner/work/pd/pd/tests/pdctl/config/config_test.go:754
2023-11-22T04:57:42.1800190Z             	            				/home/runner/work/pd/pd/tests/testutil.go:277
2023-11-22T04:57:42.1800736Z             	            				/home/runner/work/pd/pd/tests/testutil.go:259
2023-11-22T04:57:42.1801427Z             	            				/home/runner/work/pd/pd/tests/pdctl/config/config_test.go:668
2023-11-22T04:57:42.1801692Z             	Error:      	Condition never satisfied
2023-11-22T04:57:42.1802082Z             	Test:       	TestConfigTestSuite/TestUpdateDefaultReplicaConfig
```


### CI link
https://github.com/tikv/pd/actions/runs/6953034173/job/18917553194

### Reason for failure (if possible)


### Anything else

--- a/errors.toml
+++ b/errors.toml
@@ -16,6 +16,11 @@ error = '''
 redirect failed
 '''
 
+["PD:apiutil:ErrRedirectToNotLeader"]
+error = '''
+redirect to not leader
+'''
+
 ["PD:autoscaling:ErrEmptyMetricsResponse"]
 error = '''
 metrics response from Prometheus is empty

--- a/pkg/errs/errno.go
+++ b/pkg/errs/errno.go
@@ -197,6 +197,8 @@ var (
 var (
 	ErrRedirect       = errors.Normalize("redirect failed", errors.RFCCodeText("PD:apiutil:ErrRedirect"))
 	ErrOptionNotExist = errors.Normalize("the option %s does not exist", errors.RFCCodeText("PD:apiutil:ErrOptionNotExist"))
+	// ErrRedirectToNotLeader is the error message for redirect to not leader.
+	ErrRedirectToNotLeader = errors.Normalize("redirect to not leader", errors.RFCCodeText("PD:apiutil:ErrRedirectToNotLeader"))
 )
 
 // grpcutil errors

--- a/pkg/utils/apiutil/apiutil.go
+++ b/pkg/utils/apiutil/apiutil.go
@@ -61,11 +61,6 @@ const (
 	// ForwardToMicroServiceHeader is used to mark the request is forwarded to micro service.
 	ForwardToMicroServiceHeader = "Forward-To-Micro-Service"
 
-	// ErrRedirectFailed is the error message for redirect failed.
-	ErrRedirectFailed = "redirect failed"
-	// ErrRedirectToNotLeader is the error message for redirect to not leader.
-	ErrRedirectToNotLeader = "redirect to not leader"
-
 	chunkSize = 4096
 )
 
@@ -459,7 +454,7 @@ func (p *customReverseProxies) ServeHTTP(w http.ResponseWriter, r *http.Request)
 		}
 		return
 	}
-	http.Error(w, ErrRedirectFailed, http.StatusInternalServerError)
+	http.Error(w, errs.ErrRedirect.FastGenByArgs().Error(), http.StatusInternalServerError)
 }
 
 // copyHeader duplicates the HTTP headers from the source `src` to the destination `dst`.

--- a/pkg/utils/apiutil/serverapi/middleware.go
+++ b/pkg/utils/apiutil/serverapi/middleware.go
@@ -162,7 +162,13 @@ func (h *redirector) matchMicroServiceRedirectRules(r *http.Request) (bool, stri
 func (h *redirector) ServeHTTP(w http.ResponseWriter, r *http.Request, next http.HandlerFunc) {
 	redirectToMicroService, targetAddr := h.matchMicroServiceRedirectRules(r)
 	allowFollowerHandle := len(r.Header.Get(apiutil.PDAllowFollowerHandleHeader)) > 0
-	if !h.s.IsClosed() && (allowFollowerHandle || h.s.GetMember().IsLeader()) && !redirectToMicroService {
+
+	if h.s.IsClosed() {
+		http.Error(w, errs.ErrServerNotStarted.FastGenByArgs().Error(), http.StatusInternalServerError)
+		return
+	}
+
+	if (allowFollowerHandle || h.s.GetMember().IsLeader()) && !redirectToMicroService {
 		next(w, r)
 		return
 	}
@@ -170,7 +176,7 @@ func (h *redirector) ServeHTTP(w http.ResponseWriter, r *http.Request, next http
 	// Prevent more than one redirection.
 	if name := r.Header.Get(apiutil.PDRedirectorHeader); len(name) != 0 {
 		log.Error("redirect but server is not leader", zap.String("from", name), zap.String("server", h.s.Name()), errs.ZapError(errs.ErrRedirect))
-		http.Error(w, apiutil.ErrRedirectToNotLeader, http.StatusInternalServerError)
+		http.Error(w, errs.ErrRedirectToNotLeader.FastGenByArgs().Error(), http.StatusInternalServerError)
 		return
 	}
 
@@ -189,7 +195,7 @@ func (h *redirector) ServeHTTP(w http.ResponseWriter, r *http.Request, next http
 	var clientUrls []string
 	if redirectToMicroService {
 		if len(targetAddr) == 0 {
-			http.Error(w, apiutil.ErrRedirectFailed, http.StatusInternalServerError)
+			http.Error(w, errs.ErrRedirect.FastGenByArgs().Error(), http.StatusInternalServerError)
 			return
 		}
 		clientUrls = append(clientUrls, targetAddr)

--- a/server/apiv2/middlewares/redirector.go
+++ b/server/apiv2/middlewares/redirector.go
@@ -30,8 +30,13 @@ import (
 func Redirector() gin.HandlerFunc {
 	return func(c *gin.Context) {
 		svr := c.MustGet(ServerContextKey).(*server.Server)
+
+		if svr.IsClosed() {
+			c.AbortWithStatusJSON(http.StatusInternalServerError, errs.ErrServerNotStarted.FastGenByArgs().Error())
+			return
+		}
 		allowFollowerHandle := len(c.Request.Header.Get(apiutil.PDAllowFollowerHandleHeader)) > 0
-		if !svr.IsClosed() && (allowFollowerHandle || svr.GetMember().IsLeader()) {
+		if allowFollowerHandle || svr.GetMember().IsLeader() {
 			c.Next()
 			return
 		}
@@ -45,12 +50,11 @@ func Redirector() gin.HandlerFunc {
 
 		c.Request.Header.Set(apiutil.PDRedirectorHeader, svr.Name())
 
-		leader := svr.GetMember().GetLeader()
-		if leader == nil {
+		if svr.GetMember().GetLeader() == nil {
 			c.AbortWithStatusJSON(http.StatusServiceUnavailable, errs.ErrLeaderNil.FastGenByArgs().Error())
 			return
 		}
-		clientUrls := leader.GetClientUrls()
+		clientUrls := svr.GetMember().GetLeader().GetClientUrls()
 		urls := make([]url.URL, 0, len(clientUrls))
 		for _, item := range clientUrls {
 			u, err := url.Parse(item)

--- a/server/apiv2/middlewares/redirector.go
+++ b/server/apiv2/middlewares/redirector.go
@@ -31,8 +31,7 @@ func Redirector() gin.HandlerFunc {
 	return func(c *gin.Context) {
 		svr := c.MustGet(ServerContextKey).(*server.Server)
 		allowFollowerHandle := len(c.Request.Header.Get(apiutil.PDAllowFollowerHandleHeader)) > 0
-		isLeader := svr.GetMember().IsLeader()
-		if !svr.IsClosed() && (allowFollowerHandle || isLeader) {
+		if !svr.IsClosed() && (allowFollowerHandle || svr.GetMember().IsLeader()) {
 			c.Next()
 			return
 		}

